### ISSUE TITLE
Akvo Flow - Keycloak integration

### ISSUE THREAD

**Issue by iperdomo** [akvo/akvo-flow#1265]
We want to use Keycloak as SSO solution for Akvo products. Akvo FLOW can be the first to migrate to this authentication service.

**Comment by janagombitova** [comment]
# Context 

User authentication in Flow currently depends on Google, thus any Flow user needs to have a Google account. Over the past years we learned that this requirement is quite the restriction to using Flow as:  
- organisations then have to create extra accounts for users, 
- cannot use organisational accounts making it harder to ensure the right people have access,
- communication to the user via email gets unnoticed as the emails registered in Flow are not the emails users actually use. 

We have decided to take away this restriction by moving to a different authentication system. 

Secondly, with the launch of Lumen, we learned that a Flow users is also Lumen users. To access Flow data with Lumen, Salim needs to have the same email registered in Lumen that he has in Flow. Having two different systems handling user login in Flow (Google) and in Lumen (KeyCloak) creates another complication for the users to use our products. Thus we decided to implement Single Sign On (SSO). 

Thirdly, the strategy change to connect Flow and Lumen confirmed our need for having SSO implemented for the two products, with the idea to eventually have one SSO for all Akvo tools in place. 

# Decisions 

After running KeyCloak ourselves for Lumen we evaluated the feasibility to continue running it on our own and expanding its use also to Flow. But we concluded that because maintaining it ourselves costs effort, we are not sure how it will be able to handle larger traffic, we are not going to use it for authorisation, it is more feasible for us to choose an SSO service provider instead, that runs on OpenID Connect. For more details check [this article](https://docs.google.com/document/d/1x9pI86AVvxXYScTAlqIXNW3XqSEU_Fips2AYRJzGVZk/edit). 

**We will eventually stop using KeyCloak and switch to another SSO solution.** 

To decommission KeyCloak, we first need to build enough of the “grouping things app” (GTA - 1st step to brining Flow and Lumen under one roof) so that Lumen does not depend on KeyCloak. Secondly, to make this change we need to decide which solution to use.

**As working on GTA will take more effort and time, we have decided to still address the issue at hand for Flow users and implement KeyCloak for Flow.**

This means, that we will:
- move Flow authentication to Keycloak, 
- support logging in with Google, as all Flow users are registered with their Google accounts and supporting Google will either allow them to continue using their user accounts or smoothen the transition to new accounts, 
- support logging in with Miscrosoft Azure, as many of our large partners (target group) use Azure to manage their accounts across all products they use internally,
- have SSO for Flow and Lumen, 
- no longer have the restriction of needing a google account to work with our products. 

With this decision, we accepted the downsides it brings: 
- extra work to move away from Keycloak for both products once GTA is in place,
- again a new login screen for Flow our users, 
- waiting to handle better user consent to our Terms and conditions once we move to another SSO solution.  

# Next steps 
This issue is to hold to discussion conclusions and questions. Implementation will be split into smaller issues. 

We have already done a PoC were Flow uses KeyCloak to do the login, using Google as an identity provider. What needs to happen is listed in this issue: https://github.com/akvo/akvo-platform/issues/119.

## Questions 
- Do we need to redesign the KeyCloak login UI to make is fit Akvo's new brand identity and better support Google and Azure buttons?
- How will adding new users work on Flow? The difference today is that for both products you need to add in a new user, but with Lumen the user gets an invite url to accept access, where with Flow she receives nothing? Plus in Lumen you only set the user email, where in Flow you also add the user 'name'.
- With Flow we remember which language the user has selected for the UI. Lumen does this too. Will this setting be remembered across the products within KeyCloak? Also, Lumen only has 3 languages now (English, Spanish and French) where Flow has 3 more (Bahasa, Portuguese and Vietnamese). How will this work when a user goes from Flow to Lumen?

**Comment by janagombitova** [comment]
Today we discussed with the team the decisions listed above and added another variable to the discussion: **Adding users to the Flow app and eventually moving it to Google Play Store**. 

Having this goal in the equation brings us back to square one to decide which approach to take. We identified the following options: 
- Option 1: Flow and Flow app to Keycloak
   - Benefits: Move away from Google login requirement, SSO with other auth systems partners use, SSO with Lumen, easy to implement on the app, users on the Flow app and app on Google Play
   - Downsides: We will need to move everyone to AuthO (or the other provider we choose) on both Flow and Flow app, on Flow app this would mean a second larger change within a year which will affect users workflow and running projects
   - Conclusion: No go.   

- Option 2: Flow and Flow app to Auth0 (or the provider we choose)
   - Benefits: Move away from Google login requirement, SSO with other auth systems partners use, users on the Flow app and app on Google Play, Flow is future ready and we do not need to work twice on login
   - Downsides: no SSO with Lumen 
   - Conclusion: to be considered as a viable option

- Option 3: Flow to Keycloak and wait with the app till we move everything to Auth0
   - Benefits: Move away from Google login requirement, SSO with other auth systems partners use, SSO with Lumen
   - Downsides: Still no users on the Flow app and no more to Google Play store, wait till we move to Auth0 may take months, will need to move Flow login to Auth0
   - Conclusion: to be considered as a viable option

**Comment by janagombitova** [comment]
Today have we made a decision: 
- moving away for KeyCloak and using a SSO provider 
- we will make Flow future ready and **not** integrate with KeyCloak (option 3) but move it directly to the SSO provider we choose because: 
   - supporting our users to use the products at scale is our main driver, moving away from Google login lock down (but still supporting it) is one of the ways to support this, and we can live without SSO with Lumen for the time being (as this is how things are set up today)
   - we will not have to redo the login work once switching 
   - we can start preparing the app to introduce users on the app and move to a single app on Google Play Store 
- when integrating the SSO provider we do want to maintain the flexibility to switch, so we will implement the SSO solution in such a way "that the code is not all over the code base" and we can easily move to another SSO option if needed or if someone runs our code they can use a login solution of their choice easily

# Next steps 
- Research existing SSO providers in the marker and based on a set of criteria decide which to go with. 
- Once the SSO solution is chosen, break down the implementation
